proof/proof.go

```diff
@@ -17,8 +17,8 @@ import (
 type Suite interface {
 	kyber.Group
 	kyber.HashFactory
```

Comment:
@bford Since XOFs can serve perfectly well as Hashes, what if we removed hashing as a primitive from kyber. Any place someone wanted to hash, they would use the XOF from their chosen suite.

Reply:
Mhhhh. I would very much like that option to happen but we need to be careful. Some uses of `Hash` expects a fixed size output. Some uses kinda "know" that `sha256` is used, so it's compatible with external implementations (and that's bad or good depending on the point of view).

Looking at the `XOF` interface, it's not crystal clear to me how to provide a fixed size output. Should the seed be of the same size of the output's size I'm expecting ?

Reply:
Indeed, it would seem to be very nice for users of XOF to be able to use any XOF instance as a hash.Hash as well; i.e., if XOF might simply inherit from the standard hash.Hash.  

Since an XOF is semantically more powerful than a standard hash (includes all the capability of a Hash and then some), the fact that hash.Hash includes methods for producing fixed-length outputs is probably not a problem.  In fact having those methods as part of the XOF interface is probably good, too, because in many uses (of XOFs as well as hashes), we're probably just going to want "an output large enough to be secure but no larger".  In other words, some clients will want to use an XOF to produce an output whose size is determined by the client, whereas other clients will want to use an XOF to produce an output whose size the client does *not* care about, except that it should be "long enough to be secure" given the properties of the underlying XOF implementation in question (which is the XOF implementation's job to determine and not the client's).

Doing this would imply that any implementation of XOF would need to include hash.Hash, including the following two hash.Hash methods, which are specific to producing fixed-length output:

        // Sum appends the current hash to b and returns the resulting slice.
        // It does not change the underlying hash state.
        Sum(b []byte) []byte

        // Size returns the number of bytes Sum will return.
        Size() int

From a specification and documentation perspective, I would suggest that the XOF interface be extended to include a small block looking something like this:

type XOF interface {
   ... (existing stuff) ...

   // Any XOF may also be used as a Hash providing non-extensible output.
   hash.Hash

   // Size returns the minimum number of bytes of output that a client should read
   // from the XOF to produce a conventional hash with the XOF's full security strength.
   Size() int

   // Sum, a method of hash.Hash, reads Size bytes from the XOF's output and
   // appends them to input b (which may be null), returning the resulting output.
   Sum(b []byte) []byte
}

Yes, I know this will make the explicit inclusion of Size() and Sum() in the XOF interface redundant, in that we would get exactly the same interface in effect with just the hash.Hash line.  But from a documentation perspective, I'm suggesting that it's useful for the inteface to have these redundant definitions precisely so that the associated comment-documentation can explain precisely what the Sum() and Size() methods mean in the context of XOF.

Make sense and seem reasonable?  Other alternate suggestions?

Reply:
I like this idea but I'm wondering if the `hash.Hash` is necessary at all in the end. If we clearly write the methods of `hash.Hash` directly in the `XOF` interface, then by Go rules, it implements the `hash.Hash` interface, and the documentation is clear what is means wrt to `XOF`.